### PR TITLE
Handle terminating pods in the node controller (6/7)

### PR DESCRIPTION
**Pull request by smarterclayton**
Changes the node controller to repeatedly retry termination to guarantee that
pods are deleted AND finish completing before we delete a node.

Only the last commit belongs to this pull.

Extracted from #9165

### CODE REVIEW

**Comment by smarterclayton** [comment]
@gmarek some structural changes here to review - i need to run the tests a few more times to be convinced this is safe (was seeing weird failures that implied that node NotReady was not persisting).  Essentially, we only delete the node after we observe all pods are deleted.  The ordering should be NotReady (no more scheduling) -> delete all pods -> wait to observe graceful termination -> observe all pods deleted -> delete node.  Won't stop races and can't fix garbage collection of pods that are missed, but is as least as comprehensive as our existing code.

**Comment by smarterclayton** [comment]
ok to test

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -123,7 +126,9 @@ func NewNodeController(
 		kubeClient:             kubeClient,
 		recorder:               recorder,
 		podEvictionTimeout:     podEvictionTimeout,
-		podEvictor:             podEvictor,
+		maximumGracePeriod:     5 * time.Minute,
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: We probably want a flag for this.

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -145,38 +150,43 @@ func (nc *NodeController) Run(period time.Duration) {
 	}, nc.nodeMonitorPeriod)
 
 	go util.Forever(func() {
-		nc.podEvictor.TryEvict(func(nodeName string) { nc.deletePods(nodeName) })
+		nc.podEvictor.Try(func(value TimedValue) (bool, time.Duration) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I'd call variable timedNodeName, or something else more descriptive than plain 'value'

File: pkg/controller/node/nodecontroller.go [hunk 3]
```diff
@@ -145,38 +150,43 @@ func (nc *NodeController) Run(period time.Duration) {
...
-			continue
-		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		nc.recorder.Eventf(&pod, "NodeControllerEviction", "Deleting Pod %s from Node %s", pod.Name, nodeID)
-		if err := nc.kubeClient.Pods(pod.Namespace).Delete(pod.Name, nil); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
-		}
-	}
+			if completed {
+				glog.V(2).Infof("All pods terminated on %s", value.Value)
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: We can probably make this V(0), as it's likely to be a rare and important event.

File: pkg/controller/node/rate_limited_queue.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// TimedValue is a value that should be processed at a designated time.
+type TimedValue struct {
+	Value string
+	Added time.Time
+	Next  time.Time
```
Comment (gmarek) [inline pkg/controller/node/rate_limited_queue.go]: I'd like names to be more descriptive. 'Added' is probably good enough (though I'd prefer 'AddedAt', 'AddTime' or sth), but 'Next' does not say anything. In this context probably 'ProcessAt' would be better. @bgrant0607 can probably come up with better names.

File: pkg/controller/node/nodecontroller.go [hunk 5]
```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+	return len(pods.Items) > 0, nil
+}
+
+// deletePods will delete all pods from master running on given node, and return true
+// if any pods were deleted.
+func (nc *NodeController) deletePods(nodeID string) (bool, error) {
+	remaining := false
+	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.OneTermEqualSelector(client.PodHost, nodeID))
+	if err != nil {
+		return remaining, err
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: You can return false, err here.

File: pkg/controller/node/nodecontroller.go [hunk 6]
```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+	if err != nil {
+		return false, err
+	}
+	return len(pods.Items) > 0, nil
+}
+
+// deletePods will delete all pods from master running on given node, and return true
+// if any pods were deleted.
+func (nc *NodeController) deletePods(nodeID string) (bool, error) {
+	remaining := false
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: `remaining` is a really bad name for this variable.

File: pkg/controller/node/nodecontroller.go [hunk 7]
```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+		if err := nc.kubeClient.Pods(pod.Namespace).Delete(pod.Name, nil); err != nil {
+			return false, err
+		}
+		remaining = true
+	}
+	return remaining, nil
+}
+
+// terminatePods will ensure all pods on the given node that are in terminating state are eventually
+// cleaned up
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Please write what is returned from this function.

File: pkg/controller/node/nodecontroller.go [hunk 8]
```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+		nc.recordNodeEvent(nodeID, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeID))
+	}
+
+	for _, pod := range pods.Items {
+		// Defensive check, also needed for tests.
+		if pod.Spec.NodeName != nodeID {
+			continue
+		}
+		// if the pod has already been deleted, ignore it
+		if pod.DeletionGracePeriodSeconds != nil {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Is this backward compatible? IIUC `DeletionGracePeriodSeconds` was introduced quite recently, and thus will be `nil` in all 'upgraded' clusters.

File: pkg/controller/node/nodecontroller.go [hunk 9]
```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+		if next < 0 {
+			next = 0
+			glog.V(2).Infof("Removing pod %v after %s grace period", pod.Name, grace)
+			nc.recordNodeEvent(nodeID, "TerminatingEvictedPod", fmt.Sprintf("Pod %s has exceeded the grace period for deletion after being evicted from Node %q and is being force killed", pod.Name, nodeID))
+			if err := nc.kubeClient.Pods(pod.Namespace).Delete(pod.Name, api.NewDeleteOptions(0)); err != nil {
+				glog.Errorf("Error completing deletion of pod %s: %v", pod.Name, err)
+				complete = false
+			}
+		} else {
+			glog.V(2).Infof("Pod %v still terminating with %s remaining", pod.Name, next)
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Maybe s/%s remaining/%s of grace period remaining/?

File: pkg/controller/node/nodecontroller.go [hunk 10]
```diff
@@ -123,7 +126,9 @@ func NewNodeController(
 		kubeClient:             kubeClient,
 		recorder:               recorder,
 		podEvictionTimeout:     podEvictionTimeout,
-		podEvictor:             podEvictor,
+		maximumGracePeriod:     5 * time.Minute,
+		podEvictor:             NewRateLimitedTimedQueue(podEvictionLimiter, false),
+		terminationEvictor:     NewRateLimitedTimedQueue(podEvictionLimiter, false),
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Do we really want to share the same RateLimiter between podeEvictor and terminationEvictor? I'm not against it, but in general I don't like sharing state if it gives nothing except coding convenience.

File: pkg/controller/node/nodecontroller.go [hunk 11]
```diff
@@ -232,7 +242,7 @@ func (nc *NodeController) monitorNodeStatus() error {
 		for node := range deleted {
 			glog.V(1).Infof("NodeController observed a Node deletion: %v", node)
 			nc.recordNodeEvent(node, "RemovingNode", fmt.Sprintf("Removing Node %v from NodeController", node))
-			nc.deleteNode(node)
+			nc.podEvictor.Add(node)
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: (Can't comment on unchanged lines): When adding a new node you probably want to remove it from the evictor(s).

**Comment by gmarek** [comment]
IIUC the goal of this PR is to handle deletion of Pods assigned to Nodes that are removed from etcd. Current problem is that we don't retry to delete Pod in case of failure (and don't handle GracePeriod correctly, but I don't think we have to). 

I understood your proposal as:
- NC notices that the Node is deleted by comparing it's internal cache with what it observed, deleted Nodes are put into the 'outer' queue
- for deleted Node in the 'outer' queue it goes through it's Pods and if there's any pod with non-nil `DeleteGracePeriod` it puts the Node into the 'inner' queue <- this happens once per Node.
- for deleted Node in the 'inner' queue it goes through it's Pods and deletes everything that's not deleted yet and GracePeriod has passed <- this is repeated until everything is removed.

I have two questions:
- why do we want to give GracePeriod for Pods running on Nodes that are already gone and there's no way to contact them through API server (we act after Node object was deleted from etcd).
- if we want to give this GracePeriod for some reason, do we need a 2-layer logic for this? Can't we make a single 'collectGarbageFromNode' function, which I believe would look very much like your current 'inner' loop.

File: pkg/controller/node/nodecontroller.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/controller/node/nodecontroller.go]: The queue was intended to be generic enough to reuse (it's useful for any
client code that wants to wait X time before retrying an item again, like
controllers if a retry is needed), could be an interface{} but that was
more changes.

On Aug 24, 2015, at 9:58 AM, Marek Grabowski <notifications@github.com>
wrote:

In pkg/controller/node/nodecontroller.go
<https://github.com/kubernetes/kubernetes/pull/12962#discussion_r37747669>:

> @@ -145,38 +150,43 @@ func (nc *NodeController) Run(period time.Duration) {
>  	}, nc.nodeMonitorPeriod)
>
>  	go util.Forever(func() {
> -		nc.podEvictor.TryEvict(func(nodeName string) { nc.deletePods(nodeName) })
> +		nc.podEvictor.Try(func(value TimedValue) (bool, time.Duration) {

I'd call variable timedNodeName, or something else more descriptive than
plain 'value'

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962/files#r37747669>.

File: pkg/controller/node/nodecontroller.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/controller/node/nodecontroller.go]: Hrm, we typically say V(2) is normal and its paired with the eviction.
I'll move them both to info.

On Aug 24, 2015, at 10:00 AM, Marek Grabowski <notifications@github.com>
wrote:

In pkg/controller/node/nodecontroller.go
<https://github.com/kubernetes/kubernetes/pull/12962#discussion_r37747808>:

> -		return err
> -	}
> -	nc.recordNodeEvent(nodeID, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeID))
> -	for _, pod := range pods.Items {
> -		// Defensive check, also needed for tests.
> -		if pod.Spec.NodeName != nodeID {
> -			continue
> -		}
> -		glog.V(2).Infof("Delete pod %v", pod.Name)
> -		nc.recorder.Eventf(&pod, "NodeControllerEviction", "Deleting Pod %s from Node %s", pod.Name, nodeID)
> -		if err := nc.kubeClient.Pods(pod.Namespace).Delete(pod.Name, nil); err != nil {
> -			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
> -		}
> -	}
> +			if completed {
> +				glog.V(2).Infof("All pods terminated on %s", value.Value)

We can probably make this V(0), as it's likely to be a rare and important
event.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962/files#r37747808>.

File: pkg/controller/node/nodecontroller.go (hunk 2 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Yes, TimedValue as a type is OK, I'm saying that the variable name can be more descriptive. At first I wasn't certain if Value stores Node name or Pod name.

File: pkg/controller/node/nodecontroller.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/controller/node/nodecontroller.go]: It's defaulted by the API so cannot be nil - it becomes 30s everywhere

On Aug 24, 2015, at 10:33 AM, Marek Grabowski <notifications@github.com>
wrote:

In pkg/controller/node/nodecontroller.go
<https://github.com/kubernetes/kubernetes/pull/12962#discussion_r37750787>:

> +	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.OneTermEqualSelector(client.PodHost, nodeID))
> +	if err != nil {
> +		return remaining, err
> +	}
> +
> +	if len(pods.Items) > 0 {
> +		nc.recordNodeEvent(nodeID, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeID))
> +	}
> +
> +	for _, pod := range pods.Items {
> +		// Defensive check, also needed for tests.
> +		if pod.Spec.NodeName != nodeID {
> +			continue
> +		}
> +		// if the pod has already been deleted, ignore it
> +		if pod.DeletionGracePeriodSeconds != nil {

Is this backward compatible? IIUC DeletionGracePeriodSeconds was introduced
quite recently, and thus will be nil in all 'upgraded' clusters.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962/files#r37750787>.

File: pkg/controller/node/nodecontroller.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/controller/node/nodecontroller.go]: I think they are effectively the same pool - the goal seems to be to
prevent the node controller from slaughtering the master during
negotiations eviction, and delete and terminate are two parts of the same
process.  Up to you, they don't seem separate to me.

On Aug 24, 2015, at 10:47 AM, Marek Grabowski <notifications@github.com>
wrote:

In pkg/controller/node/nodecontroller.go
<https://github.com/kubernetes/kubernetes/pull/12962#discussion_r37752204>:

> @@ -123,7 +126,9 @@ func NewNodeController(
>  		kubeClient:             kubeClient,
>  		recorder:               recorder,
>  		podEvictionTimeout:     podEvictionTimeout,
> -		podEvictor:             podEvictor,
> +		maximumGracePeriod:     5 * time.Minute,
> +		podEvictor:             NewRateLimitedTimedQueue(podEvictionLimiter, false),
> +		terminationEvictor:     NewRateLimitedTimedQueue(podEvictionLimiter, false),

Do we really want to share the same RateLimiter between podeEvictor and
terminationEvictor? I'm not against it, but in general I don't like sharing
state if it gives nothing except coding convenience.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962/files#r37752204>.

**Comment by smarterclayton** [comment]
On Aug 24, 2015, at 12:01 PM, Marek Grabowski <notifications@github.com>
wrote:

IIUC the goal of this PR is to handle deletion of Pods assigned to Nodes
that are removed from etcd. Current problem is that we don't retry to
delete Pod in case of failure (and don't handle GracePeriod correctly, but
I don't think we have to).

I understood your proposal as:

   - NC notices that the Node is deleted by comparing it's internal cache
   with what it observed, deleted Nodes are put into the 'outer' queue
   - for deleted Node in the 'outer' queue it goes through it's Pods and if
   there's any pod with non-nil DeleteGracePeriod it puts the Node into the
   'inner' queue <- this happens once per Node.
   - for deleted Node in the 'inner' queue it goes through it's Pods and
   deletes everything that's not deleted yet and GracePeriod has passed <-
   this is repeated until everything is removed.

I have two questions:

   - why do we want to give GracePeriod for Pods running on Nodes that are
   already gone and there's no way to contact them through API server (we act
   after Node object was deleted from etcd).



We no longer delete first - we only delete after everything is cleaned up.
This preserves the guarantee that all pods are removed prior to deletion.

Also, the node may be severed from the master in a single direction, and I
would prefer to err on the side of letting it manage deletion on its side
before force deleting.



   - if we want to give this GracePeriod for some reason, do we need a
   2-layer logic for this? Can't we make a single 'collectGarbageFromNode'
   function, which I believe would look very much like your current 'inner'
   loop.



It was mostly to make it easier to reason about the start time - the
countdown to force deletion of pods begins when all pods have been deleted,
not when we have our first deleted pod.  The code seemed cleaner this way
(two clear phases).  Also, termination is non destructive, while deletion
is destructive.

Not a huge distinction but I was more concerned about the one way view
above (node sees evictions, but can't update status, don't want to nuke
processes).

If you think the simplification isn't worth it I'll come back to it.




—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962#issuecomment-134235005>
.

**Comment by gmarek** [comment]
Right, you're doing it also for 'normal' evictions, so we need to respect grace period.

As for the second thing - IMO adding second layer of logic just to make it easier to reason about the start time is not worth the code complexity of two layers. I guess I'm just used to slightly wider logical units. If you want to keep two layered implementation that's OK, but please make it a single call from the NC perspective (get-rid-of-assigned-pods or sth) - I don't think that leaking details of garbage collection to the 'main' NC implementation gives us anything.

**Comment by smarterclayton** [comment]
Sure.

On Tue, Aug 25, 2015 at 4:45 AM, Marek Grabowski <notifications@github.com>
wrote:

> Right, you're doing it also for 'normal' evictions, so we need to respect
> grace period.
>
> As for the second thing - IMO adding second layer of logic just to make it
> easier to reason about the start time is not worth the code complexity of
> two layers. I guess I'm just used to slightly wider logical units. If you
> want to keep two layered implementation that's OK, but please make it a
> single call from the NC perspective (get-rid-of-assigned-pods or sth) - I
> don't think that leaking details of garbage collection to the 'main' NC
> implementation gives us anything.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12962#issuecomment-134512570>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
Comments addressed in a new commit.  I realized there was also a race with
the queue.Try - if someone Remove()d while Try was processing the item, the
Remove() was ignored.  Fixed that so that a Remove() will prevent a retry.

Made the renames, added comments, isolated access to the podEvictor.

On Tue, Aug 25, 2015 at 9:16 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Sure.
>
> On Tue, Aug 25, 2015 at 4:45 AM, Marek Grabowski <notifications@github.com
> > wrote:
>
>> Right, you're doing it also for 'normal' evictions, so we need to respect
>> grace period.
>>
>> As for the second thing - IMO adding second layer of logic just to make
>> it easier to reason about the start time is not worth the code complexity
>> of two layers. I guess I'm just used to slightly wider logical units. If
>> you want to keep two layered implementation that's OK, but please make it a
>> single call from the NC perspective (get-rid-of-assigned-pods or sth) - I
>> don't think that leaking details of garbage collection to the 'main' NC
>> implementation gives us anything.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/12962#issuecomment-134512570>
>> .
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by gmarek** [comment]
LGTM. I have a thought that maybe we should explicitly make node Unschedulable before starting the eviction? I'm not sure it's currently needed, as eviction only happens when the Node is unhealthy (thus it's unchedulable), but it may be worth to do it now to be future proof. I'm leaving it up to you - if you think it's worth doing now, please do, otherwise apply lgtm label.

**Comment by smarterclayton** [comment]
Hrm - I went back and forth - I was concerned it would complicate deletion.  Let me look at the complexity (and what to do in failure).

**Comment by smarterclayton** [comment]
Marking unschedulable isn't cleanly factored now and it didn't look clean
to add, but I opened a follow up issue as
https://github.com/kubernetes/kubernetes/issues/13145 to track doing it
(unless we go with the garbage collector controller in which case it may
not be necessary).

On Tue, Aug 25, 2015 at 1:22 PM, Marek Grabowski <notifications@github.com>
wrote:

> LGTM. I have a thought that maybe we should explicitly make node
> Unschedulable before starting the eviction? I'm not sure it's currently
> needed, as eviction only happens when the Node is unhealthy (thus it's
> unchedulable), but it may be worth to do it now to be future proof. I'm
> leaving it up to you - if you think it's worth doing now, please do,
> otherwise apply lgtm label.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12962#issuecomment-134658215>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
Spawned #13145

**Comment by smarterclayton** [comment]
All green, adding label as per gmarek.

**Pull request merged**